app/uk/gov/hmrc/incorporatedentityidentification/httpparsers/GetCtReferenceHttpParser.scala

```diff
@@ -18,28 +18,34 @@ package uk.gov.hmrc.incorporatedentityidentification.httpparsers
 
 import play.api.http.Status.{NOT_FOUND, OK}
 import play.api.libs.json.{JsError, JsSuccess}
-import uk.gov.hmrc.http.{HttpReads, HttpResponse, InternalServerException}
+import uk.gov.hmrc.http.{BadGatewayException, HttpException, HttpReads, HttpResponse, NotFoundException}
+
+import scala.util.{Failure, Success, Try}
 
 object GetCtReferenceHttpParser {
 
   val ctutrKey = "CTUTR"
 
-  implicit object GetCtReferenceHttpReads extends HttpReads[Option[String]] {
-    override def read(method: String, url: String, response: HttpResponse): Option[String] = {
+  implicit object GetCtReferenceHttpReads extends HttpReads[Either[HttpException, String]] {
+    override def read(method: String, url: String, response: HttpResponse): Either[HttpException, String] = {
       response.status match {
         case OK =>
-          (response.json \ ctutrKey).validate[String] match {
-            case JsSuccess(ctutr, _) =>
-              Some(ctutr)
-            case JsError(errors) =>
-              throw new InternalServerException(s"Get CT Reference returned malformed JSON with the following errors: $errors")
+          Try(response.json) match {
+            case Success(json) =>
+              (json \ ctutrKey).validate[String] match {
+                case JsSuccess(ctutr, _) =>
+                  Right(ctutr)
+                case JsError(errors) =>
+                  Left(new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: $errors"))
+              }
+            case Failure(exception) =>
+              Left(new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: ${exception.getMessage.take(40)}"))
           }
```

Comment:
Lines 33-42 can be distilled down to:
```scala
val result = for {
  json <- Try(response.json).toEither.left.map(_.getMessage take 40)
  ctutr <- (json \ ctutrKey).validate[String].asEither.left.map(_.toString)
} yield ctutr

result.left.map { errors =>
  new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: $errors")
}
```

Reply:
https://github.com/hmrc/incorporated-entity-identification/commit/8f360d7f92d41722310f374c3b8e08863af5579d